src/script/interpreter.cpp

```diff
@@ -1781,6 +1781,35 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
     return true;
 }
 
+template <class T>
+bool GenericTransactionSignatureChecker<T>::CheckTaprootCommitment(
+    const std::vector<unsigned char>& control,
+    const std::vector<unsigned char>& program,
+    const uint256& tapleaf_hash) const
+{
+    assert(control.size() >= TAPROOT_CONTROL_BASE_SIZE);
+    assert(program.size() >= uint256::size());
+    //! The internal pubkey (x-only, so no Y coordinate parity).
+    const XOnlyPubKey p{Span{control}.subspan(1, TAPROOT_CONTROL_BASE_SIZE - 1)};
+    //! The output pubkey (taken from the scriptPubKey).
+    const XOnlyPubKey q{program};
+    // Compute the Merkle root from the leaf and the provided path.
+    const uint256 merkle_root = ComputeTaprootMerkleRoot(control, tapleaf_hash);
+    // Verify that the output pubkey matches the tweaked internal pubkey, after correcting for parity.
+    return q.CheckTapTweak(p, merkle_root, control[0] & 1);
+}
+
+template <class T>
+bool GenericTransactionSignatureChecker<T>::CheckWitnessScriptHash(
+    Span<const unsigned char> program,
+    const CScript& exec_script) const
+{
+    assert(program.size() >= uint256::size());
```

Comment:
In 70ec9f20341a5337372597ed481b048e40982b55: Instead of asserting that the program size is greater or equal to uint256 size, could we assert that the program size is exactly the `WITNESS_V0_SCRIPTHASH_SIZE`?

Reply:
I mostly added that assertion to prevent a buffer overflow in the `memcmp` below, which won't happen as long as the `program` span is 32 bytes in size or larger.

I can see the value in asserting that `program` is an actually a v0 script hash (i.e. 32 bytes exactly). I'll consider changing it if I retouch.